packages/py-moose-lib/moose_lib/main.py

```diff
@@ -2,47 +2,82 @@
 from dataclasses import dataclass, asdict
 from enum import Enum
 from string import Formatter
-from typing import Callable, Generic, Optional, TypeVar, Union
+from typing import Callable, Generic, Optional, TypeVar, Union, overload, Any
 import sys
+import os
 import json
 
-type Key[T: (str, int)] = T 
+type Key[T: (str, int)] = T
+
 
 @dataclass
 class StreamingFunction:
     run: Callable
 
+
 class IngestionFormat(Enum):
     JSON = "JSON"
     JSON_ARRAY = "JSON_ARRAY"
 
+
 @dataclass
 class IngestionConfig:
     format: Optional[IngestionFormat] = None
 
+
 @dataclass
 class StorageConfig:
     enabled: Optional[bool] = None
     order_by_fields: Optional[list[str]] = None
 
+
 @dataclass
 class DataModelConfig:
     ingestion: Optional[IngestionConfig] = None
     storage: Optional[StorageConfig] = None
 
+
 class CustomEncoder(json.JSONEncoder):
     def default(self, obj):
         if isinstance(obj, Enum):
             return obj.value
         return super().default(obj)
-    
-def moose_data_model(dataclass: type, config: Optional[DataModelConfig] = None):
-    if config:
-        config_dict = asdict(config)
-        config_json = json.dumps(config_dict, cls=CustomEncoder, indent=4)
-        print(f'{{"name": "{dataclass.__name__}Config", "config": {config_json}}}___DATAMODELCONFIG___')
-    else:
-        print("{}")
+
+
+_DC = TypeVar("_DC", bound=type)
+
+
+@overload
+def moose_data_model(arg: Optional[DataModelConfig]) -> Callable[[_DC], _DC]:
+    ...
+
+
+@overload
+def moose_data_model(arg: _DC) -> _DC:
+    ...
+
+
+def moose_data_model(arg: Any = None) -> Any:
+    def get_file(t: type) -> Optional[str]:
+        module = sys.modules.get(t.__module__)
+        if module and hasattr(module, '__file__'):
+            return module.__file__
+        return None
+
+    def decorator(data_class: type) -> type:
+        expected_file_name = os.environ.get("MOOSE_PYTHON_DM_DUMP")
```

Comment:
Check trigger env var so that other usages if data model files won't have weird side effects.